src/dex/ib-amm/ib-amm.ts

```diff
@@ -0,0 +1,215 @@
+import { Interface } from '@ethersproject/abi';
+import {
+  Token,
+  Address,
+  ExchangePrices,
+  AdapterExchangeParam,
+  SimpleExchangeParam,
+  PoolLiquidity,
+  Logger,
+} from '../../types';
+import { SwapSide, Network, ProviderURL } from '../../constants';
+import { getDexKeysWithNetwork } from '../../utils';
+import { IDex } from '../../dex/idex';
+import { IDexHelper } from '../../dex-helper/idex-helper';
+import { IbAmmData, IbAmmFunctions, IbAmmParams, PoolState } from './types';
+import { SimpleExchange } from '../simple-exchange';
+import { IbAmmConfig, Adapters } from './config';
+import IBAmmRouterABI from '../../abi/ib-amm/ib-amm.json';
+import { toLC } from './utils';
+import { JsonRpcProvider } from '@ethersproject/providers';
+import { Contract } from '@ethersproject/contracts';
+
+export class IbAmm extends SimpleExchange implements IDex<IbAmmData> {
+  readonly hasConstantPriceLargeAmounts = false;
+
+  static dexKeys = ['ibamm'];
+  public static dexKeysWithNetwork: { key: string; networks: Network[] }[] =
+    getDexKeysWithNetwork(IbAmmConfig);
+
+  logger: Logger;
+  exchangeRouterInterface: Interface;
+  poolIdentifier: string;
+
+  constructor(
+    protected network: Network,
+    protected dexKey: string,
+    protected dexHelper: IDexHelper,
+    protected adapters = Adapters[network],
+    protected config = IbAmmConfig[dexKey][network],
+  ) {
+    super(dexHelper.augustusAddress, dexHelper.provider);
+    this.logger = dexHelper.getLogger(dexKey);
+    this.exchangeRouterInterface = new Interface(IBAmmRouterABI);
+    this.poolIdentifier = `${this.dexKey}_${this.config.IBAMM_ADDRESS}`;
+  }
+
+  private poolExists(from: Token, to: Token): boolean {
+    const isBuy = toLC(from.address) === toLC(this.config.DAI);
+
+    if (toLC(from.address) === toLC(to.address)) return false;
+
+    if (isBuy) {
+      if (this.config.IB_TOKENS.every(a => toLC(a) !== toLC(to.address)))
+        return false;
+    } else {
+      if (toLC(to.address) !== toLC(this.config.MIM)) return false;
+
+      if (this.config.IB_TOKENS.every(a => toLC(a) !== toLC(from.address)))
+        return false;
+    }
+
+    return true;
+  }
+
+  private getQuote(isBuy: boolean) {
+    const provider = new JsonRpcProvider(ProviderURL[Network.MAINNET]);
```

Comment:
You should not create a provider on every quote call! You should reuse the provider available in the dexHelper